### PR TITLE
Implemented executing commands directly from console

### PR DESCRIPTION
**Pull request by roxxel**
### Summary
- Implemented executing commands directly from console
- Added ability to limit command issuing (#83)
- Small improvement of TPS counter

### CODE REVIEW

File: Obsidian/ConsoleClient.cs [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+        public bool Sleeping { get => default; set => _ = value; }
+        public bool Sneaking { get => default; set => _ = value; }
+        public bool Sprinting { get => default; set => _ = value; }
+        public bool FlyingWithElytra { get => default; set => _ = value; }
+        public bool InHorseInventory { get => default; set => _ = value; }
+        public short AttackTime { get => default; set => _ = value; }
+        public short DeathTime { get => default; set => _ = value; }
+        public short HurtTime { get => default; set => _ = value; }
+        public short SleepTimer { get => default; set => _ = value; }
+        public short CurrentSlot { get => default; set => _ = value; }
```
Comment (Seb-stian) [inline Obsidian/ConsoleClient.cs]: This looks very odd to me. Perhaps trying to access these properties (and methods) should throw `NotSupportedException` instead?

File: Obsidian/Program.cs [hunk 2]
```diff
@@ -109,6 +109,87 @@ private static async Task Main()
...
+        {
+            new Thread(new ThreadStart(async () =>
+            {
+                try
+                {
+                    Server currentServer = Servers.First().Value;
+                    Thread.Sleep(2000);
+                    while (!shutdownPending)
+                    {
+                        if (currentServer == null && Servers.Count == 0 && !shutdownPending)
```
Comment (Seb-stian) [inline Obsidian/Program.cs]: `&& !shutdownPending` looks redundant. Also, wouldn't `||` make more sense than `&&`?

File: Obsidian/Program.cs [hunk 3]
```diff
@@ -109,6 +109,87 @@ private static async Task Main()
             }
         }
 
+        private static void InitConsoleInput()
+        {
+            new Thread(new ThreadStart(async () =>
```
Comment (Seb-stian) [inline Obsidian/Program.cs]: Why is it not just a Task?

File: Obsidian/Program.cs [hunk 4]
```diff
@@ -109,6 +109,87 @@ private static async Task Main()
...
+                            {
+                                Servers.Remove(currentServer.Id);
+                                currentServer = Servers.FirstOrDefault().Value;
+                            }
+                        }
+
+                    }
+                }
+                catch (IOException) { }
+                catch (NullReferenceException) { }
```
Comment (Seb-stian) [inline Obsidian/Program.cs]: This tells me, that something in there is going wrong 😄

File: Obsidian/Server.cs [hunk 5]
```diff
@@ -290,6 +289,45 @@ public async Task StartServerAsync()
...
+                Logger.LogError($"No such command was found.");
+            }
+            catch (NoSuchParserException)
+            {
+                Logger.LogError($"The command you executed has a argument that has no matching parser.");
+            }
+            catch (InvalidCommandOverloadException)
+            {
+                Logger.LogError($"No such overload is available for this command.");
+            }
```
Comment (Seb-stian) [inline Obsidian/Server.cs]: I looked through `CommandHandler.ProcessComand` and whenever an exception is thrown, it always seems to set the `Exception.Message`, why aren't you using that?

File: Obsidian/Server.cs [hunk 6]
```diff
@@ -559,7 +597,18 @@ private async Task ServerLoop()
 
             short itersPerSecond = 0;
             var stopWatch = Stopwatch.StartNew(); // for TPS measuring
-
+            _ = Task.Run(() =>
+            {
+                while (true)
+                {
+                    if (stopWatch.ElapsedMilliseconds >= 1000L)
```
Comment (Seb-stian) [inline Obsidian/Server.cs]: I'm not sure how good idea this is. My proposal would be to replace original code with something like
```cs
TPS = (short)(1.0 / stopwatch.Elapsed.TotalSeconds);
stopwatch.Restart();
```
which would remove the need of tracking how many iterations happened in a second (`itersPerSecond`) and would update the `TPS` value every tick, instead of every second.

File: Obsidian/Utilities/ConsoleIO.cs [hunk 7]
```diff
@@ -0,0 +1,380 @@
...
+using System.Threading;
+
+namespace Obsidian.Utilities
+{
+
+    /// <summary>
+    /// Allows simultaneous console input and output without breaking user input
+    /// (Without having this annoying behaviour : User inp[Some Console output]ut)
+    /// Provide some fancy features such as formatted output, text pasting and tab-completion.
+    /// By ORelio - (c) 2012-2018 - Available under the CDDL-1.0 license
```
Comment (Seb-stian) [inline Obsidian/Utilities/ConsoleIO.cs:14]: I'm just wondering, is the license OK? Honestly, I'm not a big fan of how the whole class is implemented, so I might eventually rewrite it to something custom anyways.

Comment (Seb-stian) [inline Obsidian/Utilities/ConsoleIO.cs:14]: I'm just wondering, is the license OK? Honestly, I'm not a big fan of how the whole class is implemented, so I might eventually rewrite it to something custom anyways. Seems fine for now.

File: Obsidian/Utilities/Extensions.Colors.cs [hunk 8]
```diff
@@ -24,7 +23,7 @@ public static void RenderColoredConsoleMessage(this string message)
                 // Print text with previous color
                 if (start != i)
                 {
-                    output.Write(message.AsSpan(start, i - start));
+                    ConsoleIO.Write(message.AsSpan(start, i - start).ToString());
```
Comment (Seb-stian) [inline Obsidian/Utilities/Extensions.Colors.cs:26]: With this change, it might as well be `Substring`. (Also line 46)

File: Obsidian/Server.cs (hunk 5 above)
Comment (roxxel) [reply Obsidian/Server.cs]: I copied this code from [here](https://github.com/ObsidianMC/Obsidian/blob/eab18e7946db785c49302046cfdbfed68d46a077/Obsidian/Server.cs#L303)

File: Obsidian/Utilities/ConsoleIO.cs (hunk 7 above)
Comment (roxxel) [reply Obsidian/Utilities/ConsoleIO.cs:14]: image
From [readme](https://github.com/ORelio/Minecraft-Console-Client#readme)

File: Obsidian/ConsoleClient.cs (hunk 1 above)
Comment (roxxel) [reply Obsidian/ConsoleClient.cs]: This will force plugin devs to use try/catch blocks in commands which is not good ig

File: Obsidian/Program.cs [hunk 9]
```diff
@@ -111,82 +111,76 @@ private static async Task Main()
 
         private static void InitConsoleInput()
         {
-            new Thread(new ThreadStart(async () =>
+            Task.Run(async () =>
             {
-                try
+                Server currentServer = Servers.First().Value;
+                Thread.Sleep(2000);
```
Comment (Seb-stian) [inline Obsidian/Program.cs]:
```cs
await Task.Delay(2000);
```

File: Obsidian/Commands/MainCommandModule.cs [hunk 10]
```diff
@@ -272,14 +280,13 @@ public async Task TeleportAsync(CommandContext Context, [Remaining] VectorF loca
         [Command("op")]
         [CommandInfo("Give operator rights to a specific player.", "/op <player>")]
         [RequirePermission]
+        [IssuerScope(CommandIssuers.Any)]
         public async Task GiveOpAsync(CommandContext Context, IPlayer player)
         {
-            var server = (Server)Context.Server;
-            var onlinePlayers = server.OnlinePlayers;
-            if (!onlinePlayers.ContainsKey(player.Uuid) || !onlinePlayers.Any(x => x.Value.Username == player.Username))
+            if (player == null)
```
Comment (Naamloos) [inline Obsidian/Commands/MainCommandModule.cs:286]: isn't player also null when issuer is the console? why have CommandIssuers.Any when consoles probably can't even execute it?

File: Obsidian/ConsoleClient.cs (hunk 1 above)
Comment (Seb-stian) [reply Obsidian/ConsoleClient.cs]: Well, they should check the value of `CommandContext.Issuer`, if they allow calling from the console. Perhaps `CommandIssuers.Client` should be the default, when `[IssuerScope]` is omitted? I'm not sure.

Comment (Naamloos) [reply Obsidian/ConsoleClient.cs]: Exceptions are fine, devs should be aware of what can be done and what not.

Comment (Tides) [reply Obsidian/ConsoleClient.cs]: Hmm this does look odd. I think there should be a new interface e.x ICommandSender for this instead of having ConsoleClient implement IPlayer ^ ^

Comment (Naamloos) [reply Obsidian/ConsoleClient.cs]: in addition to that, returning `default` may return default values, giving a developer (or their code) a bad impression of what data is actually there. Just throw Exceptions like seb said

Comment (Naamloos) [reply Obsidian/ConsoleClient.cs]: Adding even more to that, it might be a good idea adding abstractions to the Client class.

File: Obsidian/Utilities/ConsoleIO.cs (hunk 7 above)
Comment (Naamloos) [reply Obsidian/Utilities/ConsoleIO.cs:14]: Good for now, but I'd prefer having this rewritten later on to avoid having weird third party licenses sprinkled in.

File: Obsidian/Utilities/Extensions.StringManipulation.cs [hunk 11]
```diff
@@ -1,5 +1,6 @@
 ﻿using Obsidian.API;
 using System;
+using System.Collections.Generic;
```
Comment (Naamloos) [inline Obsidian/Utilities/Extensions.StringManipulation.cs]: ...? Why did you add a single reference and nothing else?

**Review by Naamloos — CHANGES_REQUESTED** [review]
LGTM, but please take another look at the comments made. Won't approve until further notice.

File: Obsidian/Commands/MainCommandModule.cs (hunk 10 above)
Comment (roxxel) [reply Obsidian/Commands/MainCommandModule.cs:286]: 🤨   `player` it's the player to whom the op, not an command issuer

File: Obsidian.API/_Enums/CommandsIssuer.cs [hunk 12]
```diff
@@ -0,0 +1,20 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Obsidian.API
+{
+    [Flags]
+    public enum CommandIssuers
```
Comment (Craftplacer) [inline Obsidian.API/_Enums/CommandsIssuer.cs]: File name does not match class name

**Review by Craftplacer — CHANGES_REQUESTED** [review]
Almost there. Also, did you check up with giant list of overrides of the ConsoleClient?

File: Obsidian.API/_Enums/CommandsIssuer.cs (hunk 12 above)
Comment (roxxel) [reply Obsidian.API/_Enums/CommandsIssuer.cs]: > Almost there. Also, did you check up with giant list of overrides of the ConsoleClient?

Oh, I forgot to delete ConsoleClient, it's no longer needed

**Comment by Craftplacer** [comment]
File names still not match, but at this point, I don't care
